### ISSUE TITLE
Question. Validating nested collections and proceeding only when everything went successfully.

### ISSUE THREAD

**Issue by buinauskas** [louthy/language-ext#647]
This is a cross post from [SO](https://stackoverflow.com/questions/58397852/gathering-validation-errors-on-nested-collection).
I've got the following DTOs:

```cs
public class DatumDto
{
    public int Year { get; set; }
    public decimal Value { get; set; }
}
```

```cs
public class UpdateDataDto
{
    public int Id { get; set; }
    public IEnumerable<DatumDto> TimeSeries { get; set; }
}
```

And the following domain classes.

```cs
public class Datum
{
    public int Year { get; }
    public decimal Value { get; }

    public static Validation<Error, Datum> Of (int year, decimal value)
    {
        // ... logic

        return new Datum(year, value)
    }
}
```

```cs
public class Data
{
    public int Id { get; }
    public IEnumerable<Datum> TimeSeries { get; }

    public Validation<Error, Data> UpdateTimeSeries(IEnumerable<Datum> timeSeries)
    {
        // ... logic
    }
}
```

I've stripped some code out to have less clutter and focus on what matters to me.

Now my use case is very simple. I need to load `Data` with `TimeSeries` entities from the database and then execute a command against them, otherwise return all errors to the client.

My command looks as follows:

```cs
public class UpdateDataCommand
{
    public Validation<Error, int> Execute(IEnumerable<UpdateDataDto> arguments)
    {
        var data = _repository.Get(arguments.Map(s => s.Id));

        // part I am struggling at
        // before making any actual changes, I have to validate whether ALL DatumDtos ar valid Datums
        // If all of them are valid within all arguments - I should proceed, otherwise gather all faulty Dtos and bring them back.

        // I've come here, but it does not compile

        var result =
            from d in data
            join arg in arguments on d.Id equals arg.Id
            from datum in arg.TimeSeries
            from validation in Domain.Datum.Of(datum.Year, datum.Value)
            select d.UpdateTimeSeries(validation);   // this does not compile
    }
}
```

How would I go about validation whether all DatumDtos are valid within each element in arguments collection and only proceeding when everything went fine and gather all errors if something has gone wrong?

I understand this might be something trivial, but it's not yet _clicking_ to me. I'd appreciate help.

**Comment by bender2k14** [comment]
> This is a cross post from SO.

Please don't simultaneously cross post.

Instead of...
```
public Validation<Error, Data> UpdateTimeSeries(IEnumerable<Datum> timeSeries)
```

...do you mean...
```
public Validation<Error, Data> UpdateTimeSeries(int id, IEnumerable<Datum> timeSeries)
```

?

**Comment by buinauskas** [comment]
@bender2k14 apologies, I wasn't sure if this is going to reach anyone on SO, given that `language-ext` tag has very few questions, so I tried my best.

I did actually mean

```cs
public Validation<Error, Data> UpdateTimeSeries(int id, IEnumerable<Datum> timeSeries)
```
Code is not entirely functional and this method updates the exact instance of `Data` class.

**Comment by louthy** [comment]
@buinauskas The code not being functional isn't particularly helping, but if the question is: How can I turn an `IEnumerable` of `Validation` results into a single `Validation` result then just return:

```c#
   return result.Sequence();
```
It will flip the `IEnumerable<Validation<Error, X>>` into a `Validation<Error, IEnumerable<X>>`.  Use `Seq<A>` instead of `IEnumerable<A>` for some performance gains also.

**Comment by bender2k14** [comment]
> ...I wasn't sure if this is going to reach anyone on SO, given that language-ext tag has very few questions...

I think a better metric is how many people are watching and who is watching the `language-ext` tag.  I think @louthy watches that tag.  I used to watch it but no longer do :'(  <--- (tear of joy)  I think you would have gotten an answer.

While @louthy typed you his reply, I typed up this code for you.  Like he said, since your example isn't functional, I am not exactly sure what you want.  Nonetheless, here is some code that uses `Sequence` and `Seq<>` just like @louthy recommended.

```
public class UpdateDataCommand
{
    public Validation<Error, IEnumerable<Data>> Validate(
        IEnumerable<UpdateDataDto> arguments
    ) => arguments
        .Map(Validate)
        .Sequence();

    public Validation<Error, Data> Validate(
        UpdateDataDto argument
    ) => argument
        .TimeSeries
        .Map(d => Datum.Of(d.Year, d.Value))
        .Sequence()
        .Bind(ds => Data.UpdateTimeSeries(argument.Id, ds));
}

public class DatumDto
{
    public int Year { get; set; }
    public decimal Value { get; set; }
}

public class UpdateDataDto
{
    public int Id { get; set; }
    public Seq<DatumDto> TimeSeries { get; set; }
}

public class Datum
{
    public int Year { get; }
    public decimal Value { get; }

    private Datum(int year, decimal value)
    {
        Year = year;
        Value = value;
    }

    public static Validation<Error, Datum> Of(int year, decimal value)
    {
        // ... logic
        return new Datum(year, value);
    }
}

public class Data
{
    public int Id { get; }
    public Seq<Datum> TimeSeries { get; }

    private Data(int id, Seq<Datum> timeSeries)
    {
        Id = id;
        TimeSeries = timeSeries;
    }

    public static Validation<Error, Data> UpdateTimeSeries(int id, Seq<Datum> timeSeries)
    {
        // ... logic
        return new Data(id, timeSeries);
    }
}
```

**Comment by buinauskas** [comment]
Heya, this actually looks very good and promising. I'll perhaps add more dummy data here to illustrate what would be the arguments and what would be the failure just to be clear we on the same page?

Invalid list. At least one Value in the list within DatumDto has a negative value. So I'd expect my Command tu return validation errors.

```cs
new List<Data>
{
    new Data
    {
        Id = 1,
        TimeSeries = new List<DatumDto>
        {
            new DatumDto { Year = 2020, Value = 100 },
            new DatumDto { Year = 2020, Value = -10 }
        }
    },
    new Data
    {
        Id = 2,
        TimeSeries = new List<DatumDto>
        {
            new DatumDto { Year = 2020, Value = 100 },
            new DatumDto { Year = 2020, Value = 210 }
        }
    }
}
```

Valid list. All values within DatumDto are non-negative. This is success.

```cs
new List<Data>
{
    new Data
    {
        Id = 1,
        TimeSeries = new List<DatumDto>
        {
            new DatumDto { Year = 2020, Value = 100 },
            new DatumDto { Year = 2020, Value = 210 }
        }
    },
    new Data
    {
        Id = 2,
        TimeSeries = new List<DatumDto>
        {
            new DatumDto { Year = 2020, Value = 100 },
            new DatumDto { Year = 2020, Value = 310 }
        }
    }
}
```

Also just because my code wasn't functional, doesn't mean it won't be ; )

**Comment by bender2k14** [comment]
> To be clear what you've done was that you first added `Validate` method that's going to validate a single `UpdateDataDto` and then use this validator as a parameter to validate entire collection, correct?

Correct.  And you can call that method whatever you want.  The name `Validate` was just the first thing I thought of.

> I'll perhaps add more dummy data here to illustrate what would be the arguments and what would be the failure just to be clear we on the same page?

Can you do this by writing tests?  If you were having a problem and could express your problem with a failing test, then the chances are very likely that we can help you :)

> > > Code is not entirely functional...
>>
> > The code not being functional isn't particularly helping
>
> Also just because my code wasn't functional, doesn't mean it won't be ; )

When @louthy used the word "functional" there, he meant "working" or "code that compiles".  So he meant something like "The code not compiling isn't particularly helping...".  We recommend that you share code with us that we can execute (like a test!).  That helps us so that we can help you.

**Comment by buinauskas** [comment]
🤦‍♂️🤦‍♂️🤦‍♂️

Alright. Ill try this out tomorrow and see how it has worked out. In any case I'll come back with a proper test case to share my own error or success.

Thank you both for help!

**Comment by buinauskas** [comment]
Apologies, haven't had time to implement the whole thing, was too busy. However I did prepare a test-case with all the code, which I'll use myself to attempt my implementation. Here is it:

validationtest.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <IsPackable>false</IsPackable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0"/>
    <PackageReference Include="xunit" Version="2.4.0"/>
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0"/>
    <PackageReference Include="coverlet.collector" Version="1.0.1"/>
    <PackageReference Include="FluentAssertions" Version="5.9.0"/>
    <PackageReference Include="LanguageExt.Core" Version="3.3.28"/>
    <PackageReference Include="LanguageExt.UnitTesting" Version="3.3.5"/>
  </ItemGroup>
</Project>
```

ValidationLogic.cs
```cs
using System;
using System.Collections.Generic;
using System.Linq;
using LanguageExt;
using LanguageExt.Common;

namespace validationtest
{
    #region  Data Transfer Objects
    public class DatumDto
    {
        public int Year { get; set; }
        public int Value { get; set; }
    }

    public class UpdatePriceHistoryDto
    {
        public int CountryId { get; set; }
        public int CategoryId { get; set; }
        public IEnumerable<DatumDto> TimeSeries { get; set; }
    }
    #endregion

    #region Business objects
    public class Datum : Record<Datum>
    {
        public int Year { get; }
        public int Value { get; }

        public Datum(int year, int value)
        {
            Year = year;
            Value = value;
        }

        public static Validation<Error, Datum> Of(int year, int value)
        {
            if (value < 0)
            {
                return Error.New($"{nameof(value)} has to be non-negative value");
            }

            return new Datum(year, value);
        }
    }
    public class PriceHistoryId : Record<PriceHistoryId>
    {
        public int CountryId { get; }
        public int CategoryId { get; }

        public PriceHistoryId(int countryId, int categoryId)
        {
            CountryId = countryId;
            CategoryId = categoryId;
        }
    }

    public class PriceHistory
    {
        private readonly List<Datum> _timeSeries;

        public PriceHistoryId Id { get; }
        public IEnumerable<int> AvailableYears { get; }
        public IEnumerable<Datum> TimeSeries => _timeSeries.ToList();

        public PriceHistory(PriceHistoryId id, List<int> availableYears, List<Datum> timeSeries)
        {
            Id = id;
            AvailableYears = availableYears;
            _timeSeries = timeSeries;
        }

        public Validation<Error, PriceHistory> UpdateTimeSeries(IEnumerable<Datum> timeSeries)
        {
            if(!timeSeries.All(datum => AvailableYears.Contains(datum.Year)))
            {
                return Error.New("New time series has to be within range of available years");
            }

            _timeSeries.RemoveAll(s => timeSeries.Select(t => t.Year).Contains(s.Year));

            _timeSeries.AddRange(timeSeries);

            return this;
        }

        #region Domain entities are equal when their ids are equal
        private object Actual => this;

        public override bool Equals(object obj)
        {
            var other = obj as PriceHistory;

            if (other is null)
                return false;

            if (ReferenceEquals(this, other))
                return true;

            if (Actual.GetType() != other.Actual.GetType())
                return false;

            return Id == other.Id;
        }

        public static bool operator ==(PriceHistory a, PriceHistory b)
        {
            if (a is null && b is null)
                return true;

            if (a is null || b is null)
                return false;

            return a.Equals(b);
        }

        public static bool operator !=(PriceHistory a, PriceHistory b) => !(a == b);

        public override int GetHashCode() => Id.GetHashCode();
        #endregion
    }
    #endregion

    #region Update Price History command
    public static class UpdatePriceHistoryTimeSeriesCommand
    {
        public static Validation<Error, PriceHistory> Handle(IEnumerable<PriceHistory> histories, IEnumerable<UpdatePriceHistoryDto> priceHistoryDtos)
        {
            throw new NotImplementedException($"{nameof(Handle)} has not been implemented");
        }
    }
    #endregion
}
```

ValidationTest.cs
```cs
using System.Collections.Generic;
using FluentAssertions;
using LanguageExt.UnitTesting;
using Xunit;

namespace validationtest
{
    public class ValidationTest
    {
        [Fact]
        public void CompareEquality_GivenTwoPriceHistoryIds_Equals()
        {
            var id1 = new PriceHistoryId(1, 1);
            var id2 = new PriceHistoryId(1, 1);

            id1.Should().BeEquivalentTo(id2);
        }

        [Fact]
        public void CompareEquality_GivenTwoPriceHistories_Equals()
        {
            var id1 = new PriceHistoryId(1, 1);
            var ay1 = new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 };
            var ts1 = new List<Datum> { new Datum(2001, 20), new Datum(2002, 20) };
            var ph1 = new PriceHistory(id1, ay1, ts1);

            var id2 = new PriceHistoryId(1, 1);
            var ay2 = new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 };
            var ts2 = new List<Datum> { new Datum(2001, 20), new Datum(2002, 20), new Datum(2003, 20), new Datum(2004, 20) };
            var ph2 = new PriceHistory(id2, ay2, ts2);

            ph1.Should().BeEquivalentTo(ph2);
        }


        [Fact]
        public void UpdateTimeSeries_GivenValidDatums_Updates()
        {
            var id = new PriceHistoryId(1, 1);
            var ay = new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 };
            var ts = new List<Datum> { new Datum(2002, 20) };
            var ph = new PriceHistory(id, ay, ts);

            var datums = new List<Datum>
            {
                new Datum(2001, 2001),
                new Datum(2002, 2002),
                new Datum(2003, 2003),
            };

            var expectedDatums = new List<Datum>
            {
                new Datum(2001, 2001),
                new Datum(2002, 2002),
                new Datum(2003, 2003),
            };

            var result = ph.UpdateTimeSeries(datums);

            result.ShouldBeSuccess(priceHistory => priceHistory.TimeSeries.Should().BeEquivalentTo(expectedDatums));
        }

        [Fact]
        public void UpdateTimeSeries_GivenOutOfRangeDatums_Updates()
        {
            var id = new PriceHistoryId(1, 1);
            var ay = new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 };
            var ts = new List<Datum> { new Datum(2000, 1), new Datum(2001, 20), new Datum(2002, 20) };
            var ph = new PriceHistory(id, ay, ts);

            var datums = new List<Datum>
            {
                new Datum(2000, 2000),
                new Datum(2001, 2001),
                new Datum(2002, 2002),
                new Datum(2003, 2003),
            };

            var result = ph.UpdateTimeSeries(datums);

            result.ShouldBeFail();
        }

        [Fact]
        public void UpdatePriceHistoryTimeSeriesCommand_GivenValidPriceHistoryDtos_UpdatesPriceHistory()
        {
            var command = new List<UpdatePriceHistoryDto>
            {
                new UpdatePriceHistoryDto
                {
                    CountryId = 1,
                    CategoryId = 1,
                    TimeSeries = new List<DatumDto>
                    {
                        new DatumDto { Year = 2001, Value = 10 },
                        new DatumDto { Year = 2002, Value = 20 },
                        new DatumDto { Year = 2003, Value = 30 },
                    }
                },
                new UpdatePriceHistoryDto
                {
                    CountryId = 2,
                    CategoryId = 2,
                    TimeSeries = new List<DatumDto>
                    {
                        new DatumDto { Year = 2001, Value = 40 },
                        new DatumDto { Year = 2002, Value = 50 },
                        new DatumDto { Year = 2003, Value = 60 },
                    }
                }
            };

            var histories = new List<PriceHistory>
            {
                new PriceHistory(
                    new PriceHistoryId(1, 1),
                    new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 },
                    new List<Datum> { new Datum(2001, 15) }
                ),
                new PriceHistory(
                    new PriceHistoryId(2, 2),
                    new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 },
                    new List<Datum> { new Datum(2004, 15) }
                )
            };

            var expectedHistories = new List<PriceHistory>
            {
                new PriceHistory(
                    new PriceHistoryId(1, 1),
                    new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 },
                    new List<Datum> { new Datum(2001, 10), new Datum(2002, 20), new Datum(2003, 30) }
                ),
                new PriceHistory(
                    new PriceHistoryId(2, 2),
                    new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 },
                    new List<Datum> { new Datum(2004, 15) }
                )
            };

            var result = UpdatePriceHistoryTimeSeriesCommand.Handle(histories, command);

            result.ShouldBeSuccess(r => r.Should().BeEquivalentTo(expectedHistories));
        }

        [Fact]
        public void UpdatePriceHistoryTimeSeriesCommand_GivenInvalidPriceHistoryDtos_ReturnsValidationError()
        {
            var command = new List<UpdatePriceHistoryDto>
            {
                new UpdatePriceHistoryDto
                {
                    CountryId = 1,
                    CategoryId = 1,
                    TimeSeries = new List<DatumDto>
                    {
                        new DatumDto { Year = 2001, Value = 10 },
                        new DatumDto { Year = 2002, Value = 20 },
                        new DatumDto { Year = 2003, Value = 30 },
                    }
                },
                new UpdatePriceHistoryDto
                {
                    CountryId = 2,
                    CategoryId = 2,
                    TimeSeries = new List<DatumDto>
                    {
                        new DatumDto { Year = 2001, Value = 40 },
                        new DatumDto { Year = 2002, Value = 50 },
                        new DatumDto { Year = 2003, Value = 60 },
                    }
                }
            };

            var histories = new List<PriceHistory>
            {
                new PriceHistory(
                    new PriceHistoryId(1, 1),
                    new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 },
                    new List<Datum> { new Datum(2001, 15) }
                ),
                new PriceHistory(
                    new PriceHistoryId(2, 2),
                    new List<int> { 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010 },
                    new List<Datum> { new Datum(2004, 15) }
                )
            };

            var result = UpdatePriceHistoryTimeSeriesCommand.Handle(histories, command);

            result.ShouldBeFail();
        }

        [Fact]
        public void UpdatePriceHistoryTimeSeriesCommand_GivenDatumDtosOutOfAvailableYearsRange_ReturnsValidationError()
        {
            var command = new List<UpdatePriceHistoryDto>
            {
                new UpdatePriceHistoryDto
                {
                    CountryId = 1,
                    CategoryId = 1,
                    TimeSeries = new List<DatumDto>
                    {
                        new DatumDto { Year = 2001, Value = 10 },
                        new DatumDto { Year = 2002, Value = 20 },
                        new DatumDto { Year = 2003, Value = 30 },
                    }
                },
                new UpdatePriceHistoryDto
                {
                    CountryId = 2,
                    CategoryId = 2,
                    TimeSeries = new List<DatumDto>
                    {
                        new DatumDto { Year = 2001, Value = 40 },
                        new DatumDto { Year = 2002, Value = 50 },
                        new DatumDto { Year = 2003, Value = 60 },
                    }
                }
            };

            var histories = new List<PriceHistory>
            {
                new PriceHistory(
                    new PriceHistoryId(1, 1),
                    new List<int> { 2001, 2002 },
                    new List<Datum> { new Datum(2001, 15) }
                ),
                new PriceHistory(
                    new PriceHistoryId(2, 2),
                    new List<int> { 2001, 2002 },
                    new List<Datum> { new Datum(2004, 15) }
                )
            };

            var result = UpdatePriceHistoryTimeSeriesCommand.Handle(histories, command);

            result.ShouldBeFail();
        }
    }
}
```

**Comment by bender2k14** [comment]
Are you asking for help with your previous comment?

**Comment by buinauskas** [comment]
@bender2k14 in previous comment you mentioned
>We recommend that you share code with us that we can execute (like a test!). That helps us so that we can help you.

So I've taken my time and wrote that code above with expected logic and a failing test. I understood that you wanted me to come better prepared and have a test that's actually a spec.

In an essence yes, I wanted some help with my previous comment and gave fully working example of what I'd like to implement and what directions should I take. Apologies if I misunderstood you.

**Comment by bender2k14** [comment]
Ok, great.  You have done an excellent job.  I just wanted to make sure I understood that comment correctly.

 > So I've taken my time and wrote that code above with expected logic and a failing test.

Just one failing test?  Do all the other tests you gave pass?

**Comment by buinauskas** [comment]
Not all of them pass. However I had enough time today and by looking at code you have written to me, I managed to implement everything I needed. Code's really short, descriptive.

This sample project with proper test cases was really helpful because I was able to solve on exact problem.

I appreciate all the help, thank you @bender2k14 and @louthy.

In case I have any questions in the future, providing a failing test case with expectations as in previous comment would work? Or is there a way I could be more clear with my questions?
